Implement IPC error string reporting in testing for IPC::Connection

jon-b-apple opened the pull request:
#### 2837e1eaeb3d3d5e7ac3bb0535c8ef1b9ee4fddb
<pre>
Implement IPC error string reporting in testing for IPC::Connection
<a href="https://rdar.apple.com/163035367">rdar://163035367</a>
<a href="https://bugs.webkit.org/show_bug.cgi?id=301095">https://bugs.webkit.org/show_bug.cgi?id=301095</a>

Reviewed by Kimmo Kinnunen.

This patch introduces the first part of a feature to enable error reporting over IPC for validation failures and MESSAGE_CHECKs. This feature currently only supports the base IPC::Connection, and not stream connections at this time.

It adds error string tracking to the IPC Decoder and Connection, and propagates these during decoding. Error strings are stored until they are read by WebContent by calling the appropriate TakeInvalidMessageStringForTesting message on the underlying IPC connection. This allows tests to trigger specific error conditions and check the resulting string to make sure the expected condition was reached.

This patch also adds the IPCTester.CheckTestParameter method and the ipc/get-error-string.html test to demonstrate the use of this feature.

This feature is only enabled for builds using the IPC Testing API.

Tests: ipc/get-error-string.html
       Tools/TestWebKitAPI/Tests/IPC/ConnectionTests.cpp
       Tools/TestWebKitAPI/Tests/IPC/StreamConnectionTests.cpp

* LayoutTests/ipc/get-error-string-expected.txt: Added.
* LayoutTests/ipc/get-error-string.html: Added.
* Source/WebKit/DerivedSources-input.xcfilelist:
* Source/WebKit/DerivedSources.make:
* Source/WebKit/GPUProcess/GPUConnectionToWebProcess.cpp:
(WebKit::GPUConnectionToWebProcess::takeInvalidMessageStringForTesting):
* Source/WebKit/GPUProcess/GPUConnectionToWebProcess.h:
* Source/WebKit/GPUProcess/GPUConnectionToWebProcess.messages.in:
* Source/WebKit/NetworkProcess/NetworkConnectionToWebProcess.cpp:
(WebKit::NetworkConnectionToWebProcess::takeInvalidMessageStringForTesting):
* Source/WebKit/NetworkProcess/NetworkConnectionToWebProcess.h:
* Source/WebKit/NetworkProcess/NetworkConnectionToWebProcess.messages.in:
* Source/WebKit/Platform/IPC/Connection.cpp:
(IPC::Connection::dispatchMessageReceiverMessage):
(IPC::Connection::dispatchMessage):
* Source/WebKit/Platform/IPC/Connection.h:
(IPC::Connection::hasErrorString const):
(IPC::Connection::setErrorString):
(IPC::Connection::takeErrorString):
(IPC::Connection::markCurrentlyDispatchedMessageAsInvalid):
(IPC::markCurrentlyDispatchedMessageAsInvalid):
* Source/WebKit/Platform/IPC/Decoder.h:
(IPC::Decoder::hasErrorString const):
(IPC::Decoder::setErrorString):
(IPC::Decoder::takeErrorString):
* Source/WebKit/Platform/IPC/StreamServerConnection.cpp:
(IPC::StreamServerConnection::markCurrentlyDispatchedMessageAsInvalid):
* Source/WebKit/Platform/IPC/StreamServerConnection.h:
(IPC::markCurrentlyDispatchedMessageAsInvalid):
* Source/WebKit/Scripts/generate-serializers.py:
(decode_type):
* Source/WebKit/Shared/API/Cocoa/RemoteObjectRegistry.mm:
(WebKit::RemoteObjectRegistry::callReplyBlock):
* Source/WebKit/Shared/IPCTester.cpp:
(WebKit::IPCTester::checkTestParameter):
* Source/WebKit/Shared/IPCTester.h:
* Source/WebKit/Shared/IPCTester.messages.in:
* Source/WebKit/Shared/IPCTester.serialization.in: Added.
* Source/WebKit/Shared/TestParameter.h: Added.
* Source/WebKit/UIProcess/WebProcessProxy.cpp:
(WebKit::WebProcessProxy::takeInvalidMessageStringForTesting):
* Source/WebKit/UIProcess/WebProcessProxy.h:
* Source/WebKit/UIProcess/WebProcessProxy.messages.in:
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Tools/TestWebKitAPI/Tests/IPC/ConnectionTests.cpp:
* Tools/TestWebKitAPI/Tests/IPC/StreamConnectionTests.cpp:

Canonical link: <a href="https://commits.webkit.org/302528@main">https://commits.webkit.org/302528@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/b66fd047a690afc660ddca19efebc29c941cc301

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/129250 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/159/builds/1506 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/40086 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/136626 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/80639 "Built successfully") | [✅ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/fffd0010-024c-45bb-83c4-8a41d6b8030a) 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/131121 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/155/builds/1439 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/156/builds/1383 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/98422 "Passed tests") | [❌ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/66322 "") | [❌ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/126c9174-106a-40b4-9cc0-946e6bb30226) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/132197 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/162/builds/1122 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/115778 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/79073 "Passed tests") | | [✅ 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/9cf5b7ee-93aa-41a5-967f-9738ee1d7fba) 
| [✅ 🧪 webkitpy](https://ews-build.webkit.org/#/builders/6/builds/128600 "Passed tests") | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/154/builds/1048 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/33897 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/79905 "Built successfully") | | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/109501 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/34392 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/139100 "Built successfully") | | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/153/builds/1297 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/161/builds/1248 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/106952 "Passed tests") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/160/builds/1347 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/112114 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/106790 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/27217 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/164/builds/1070 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/30634 "Passed tests") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/53913 "The change is no longer eligible for processing.") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/158/builds/1369 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/64725 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/157/builds/1192 "Built successfully") | | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/163/builds/1229 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/152/builds/1292 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

sheeparegreat commented on Source/WebKit/Platform/IPC/Decoder.h:
```diff
@@ -113,6 +113,16 @@ class Decoder {
+        if (!hasErrorString())
+            m_errorString = error;
+    }
+    const char* takeErrorString() { const char* res = m_errorString; m_errorString = nullptr; return res; }
```
std::exchange?

achristensen07 commented on Source/WebKit/Platform/IPC/Decoder.h:
```diff
@@ -205,6 +215,10 @@ class Decoder {
     Vector<uint32_t> m_indicesOfObjectsFailingDecoding;
+
+#if ENABLE(IPC_TESTING_API)
+    const char* m_errorString { nullptr };
```
Let's use ASCIILiteral to make it more clear that this doesn't need lifetime management.

achristensen07 commented on Source/WebKit/Platform/IPC/Connection.h:
```diff
@@ -29,6 +29,7 @@
 #pragma once
 
 #include "ConnectionHandle.h"
+#include "MessageNames.h"
```
Can we put inline functions in Connection.cpp and not include new headers from this widely-included header?

kkinnunen-apple commented on Source/WebKit/Platform/IPC/Connection.h:
```diff
@@ -1106,5 +1127,34 @@ inline void markCurrentlyDispatchedMessageAsInvalid(const RefPtr<Connection>& co
+    connection->markCurrentlyDispatchedMessageAsInvalid();
+}
+
+#if ENABLE(IPC_TESTING_API)
```
To prevent proliferation of these functions, could you perhaps just add the string as a parameter to markCurrentlyDispatchedMessageAsInvalid?

kkinnunen-apple commented on LayoutTests/ipc/get-error-string.html:
```diff
@@ -0,0 +1,49 @@
+
+    // Test case 2: Boundary value that should pass both validations (99 > 10 and 99 < 100)
+    CoreIPC.GPU.IPCTester.CheckTestParameter(0, { param: { value: 99 } });
+    errorString = IPC.getErrorString("GPU");
```
The consistency of this API addition is strange.
The message is sent via `CoreIPC.GPU.IPCTester.CheckTestParameter`
The error check message is sent via `IPC.getErrorString("GPU")`

In other words, you use higher level abstraction to send the message, but then you escape to implementation detail, e.g. lower level abstraction, to check the error state.

I think the most consistent would be:
```
    CoreIPC.GPU.IPCTester.CheckTestParameter(0, { param: { value: 10 } });
    let errorString = CoreIPC.GPU.GPUConnectionToWebProcess.TakeInvalidMessageStringForTesting(0);
```
```
    CoreIPC.Network.IPCTester.CheckTestParameter(0, { param: { value: 10 } });
    let errorString = CoreIPC.Networking.NetworkConnectionToWebProcessTakeInvalidMessageStringForTesting(0);
```

or extra function in coreipc.js

```
    CoreIPC.GPU.IPCTester.CheckTestParameter(0, { param: { value: 10 } });
    let errorString = COreIPC.GPU.takeErrorString();
```

kkinnunen-apple commented on Source/WebKit/WebProcess/WebPage/IPCTestingAPI.cpp:
```diff
@@ -2742,6 +2744,47 @@ JSValueRef JSIPC::removeTesterReceiver(JSContextRef context, JSObjectRef, JSObje
     return JSValueMakeUndefined(context);
 }
 
+JSValueRef JSIPC::getErrorString(JSContextRef context, JSObjectRef, JSObjectRef thisObject, size_t argumentCount, const JSValueRef arguments[], JSValueRef* exception)
```
This could be perhaps `takeErrorString`. 
If it is a getter, it could be `errorString` but it doesn't seem to be a getter.

kkinnunen-apple commented on Source/WebKit/Scripts/generate-serializers.py:
```diff
@@ -966,8 +966,12 @@ def decode_type(type, serialized_types):
-                result.append(f'    if (!({validator}))')
+                result.append(f'    if (!({validator})) {{')
+                result.append('#if ENABLE(IPC_TESTING_API)')
+                result.append(f'        decoder.setErrorString("Validation failed: {validator}");')
```
Is this missing the `_s`?


Also if I understand correctly, the effect of this change should be visible in the generated example files in tests/, but it isn't. Maybe the test content generator wasn't run?
(`make -C Source/WebKit/Scripts/webkit/tests` or similar?)

kkinnunen-apple commented on Source/WebKit/Platform/IPC/HandleMessage.h:
```diff
@@ -321,8 +331,13 @@ void handleMessage(C& connection, Decoder& decoder, T* object, MF U::* function)
+    if (!arguments) [[unlikely]] {
+#if ENABLE(IPC_TESTING_API)
+        if constexpr (std::is_same_v<C, IPC::Connection>)
+            setConnectionErrorStringIfNeeded(connection, decoder);
```
This should not be here, not all messages are processed with handle message.
You can see Connection.cpp, search for m_ignoreInvalidMessageForTesting. There are the locations roughly where you need to convert message from decoder into message from connection.

In addition of missing cases, the problem is that we have thousands of handle locations, but this can be covered in the single location. So no need to spend the binary size when this can be implemented at the call site in Connection.cpp.

kkinnunen-apple reviewed (changes requested):
You should improve the consistency of propagating the error in all cases.
The two cases that cause connection to be invalid:
1. decoder error (returning nullopt  from decoder decode).
2. validation error (calling Connection::markCurrentlyDispatchedMessageAsInvalid).

From a quick glance, case 1. seems to be handled ok .
From a quick glance, case 2. fails to address a lot of cases.

The suggestion is:
Instead of Connection::setConnectionErrorString, use the markCurrentlyDispatchedMessageAsInvalid

```
void Connection:: markCurrentlyDispatchedMessageAsInvalid(ASCIILiteral error)
```

Consider also structural change, in case you have energy:
Instead of adding IPC::Connection -wide protocol message (TakeInvalidMessageErrorString), add GPUConnectionToWebProcess, NetworkConnectionToWebProcess, WebProcessProxy messages TakeInvalidMessageErrorString.
This way the impact would be more minimal, as the logic would not pollute the whole IPC Connection  domain, rather the few tested clients.

kkinnunen-apple commented on Source/WebKit/Platform/IPC/Connection.h:
```diff
@@ -803,6 +819,11 @@ class Connection : public ThreadSafeRefCountedAndCanMakeThreadSafeWeakPtr<Connec
 #endif
+
+#if ENABLE(IPC_TESTING_API)
+    ASCIILiteral m_errorString { nullptr };
```
nit: ASCIILiteral default constructor uses the nullptr..

kkinnunen-apple commented on Source/WebKit/Platform/IPC/Connection.h:
```diff
@@ -551,6 +557,16 @@ class Connection : public ThreadSafeRefCountedAndCanMakeThreadSafeWeakPtr<Connec
     static void setShouldCrashOnMessageCheckFailure(bool);
 
+#if ENABLE(IPC_TESTING_API)
+    bool hasErrorString() const { return !m_errorString.isNull(); }
```
nit: probably the hasErrorString is a redundant function..

kkinnunen-apple commented on Source/WebKit/Platform/IPC/Connection.cpp:
```diff
@@ -529,6 +532,11 @@ Error Connection::flushSentMessages(Timeout timeout)
     m_isValid = false;
+
+#if ENABLE(IPC_TESTING_API)
+    setErrorString({ });
```
I think it might be useful to keep the error state if invalidate is called, and no downside as far as I see.
An error handling codepath might run common code that runs invalidate(), which would then lose the error information that might be accessed by a test that uses the feature to test the error condition.

kkinnunen-apple reviewed (approved):
Nice, much simpler!

jon-b-apple committed:
Implement IPC error string reporting in testing for IPC::Connection
rdar://163035367
https://bugs.webkit.org/show_bug.cgi?id=301095

Reviewed by Kimmo Kinnunen.

This patch introduces the first part of a feature to enable error reporting over IPC for validation failures and MESSAGE_CHECKs. This feature currently only supports the base IPC::Connection, and not stream connections at this time.

It adds error string tracking to the IPC Decoder and Connection, and propagates these during decoding. Error strings are stored until they are read by WebContent by calling the appropriate TakeInvalidMessageStringForTesting message on the underlying IPC connection. This allows tests to trigger specific error conditions and check the resulting string to make sure the expected condition was reached.

This patch also adds the IPCTester.CheckTestParameter method and the ipc/get-error-string.html test to demonstrate the use of this feature.

This feature is only enabled for builds using the IPC Testing API.

Tests: ipc/get-error-string.html
       Tools/TestWebKitAPI/Tests/IPC/ConnectionTests.cpp
       Tools/TestWebKitAPI/Tests/IPC/StreamConnectionTests.cpp

* LayoutTests/ipc/get-error-string-expected.txt: Added.
* LayoutTests/ipc/get-error-string.html: Added.
* Source/WebKit/DerivedSources-input.xcfilelist:
* Source/WebKit/DerivedSources.make:
* Source/WebKit/GPUProcess/GPUConnectionToWebProcess.cpp:
(WebKit::GPUConnectionToWebProcess::takeInvalidMessageStringForTesting):
* Source/WebKit/GPUProcess/GPUConnectionToWebProcess.h:
* Source/WebKit/GPUProcess/GPUConnectionToWebProcess.messages.in:
* Source/WebKit/NetworkProcess/NetworkConnectionToWebProcess.cpp:
(WebKit::NetworkConnectionToWebProcess::takeInvalidMessageStringForTesting):
* Source/WebKit/NetworkProcess/NetworkConnectionToWebProcess.h:
* Source/WebKit/NetworkProcess/NetworkConnectionToWebProcess.messages.in:
* Source/WebKit/Platform/IPC/Connection.cpp:
(IPC::Connection::dispatchMessageReceiverMessage):
(IPC::Connection::dispatchMessage):
* Source/WebKit/Platform/IPC/Connection.h:
(IPC::Connection::hasErrorString const):
(IPC::Connection::setErrorString):
(IPC::Connection::takeErrorString):
(IPC::Connection::markCurrentlyDispatchedMessageAsInvalid):
(IPC::markCurrentlyDispatchedMessageAsInvalid):
* Source/WebKit/Platform/IPC/Decoder.h:
(IPC::Decoder::hasErrorString const):
(IPC::Decoder::setErrorString):
(IPC::Decoder::takeErrorString):
* Source/WebKit/Platform/IPC/StreamServerConnection.cpp:
(IPC::StreamServerConnection::markCurrentlyDispatchedMessageAsInvalid):
* Source/WebKit/Platform/IPC/StreamServerConnection.h:
(IPC::markCurrentlyDispatchedMessageAsInvalid):
* Source/WebKit/Scripts/generate-serializers.py:
(decode_type):
* Source/WebKit/Shared/API/Cocoa/RemoteObjectRegistry.mm:
(WebKit::RemoteObjectRegistry::callReplyBlock):
* Source/WebKit/Shared/IPCTester.cpp:
(WebKit::IPCTester::checkTestParameter):
* Source/WebKit/Shared/IPCTester.h:
* Source/WebKit/Shared/IPCTester.messages.in:
* Source/WebKit/Shared/IPCTester.serialization.in: Added.
* Source/WebKit/Shared/TestParameter.h: Added.
* Source/WebKit/UIProcess/WebProcessProxy.cpp:
(WebKit::WebProcessProxy::takeInvalidMessageStringForTesting):
* Source/WebKit/UIProcess/WebProcessProxy.h:
* Source/WebKit/UIProcess/WebProcessProxy.messages.in:
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Tools/TestWebKitAPI/Tests/IPC/ConnectionTests.cpp:
* Tools/TestWebKitAPI/Tests/IPC/StreamConnectionTests.cpp:

Canonical link: https://commits.webkit.org/302528@main
